### PR TITLE
Quantum ESPRESSO update - Nov 2018

### PR DESCRIPTION
**Pull request by naromero77**
Several enhancements to the Quantum ESPRESSO Spack package:
- Proper detection and use of BLAS and LAPACK provided by Spack.
- Proper detection and use of FFTs via new FFTW-API provider.
- Use releases from gitlab.com instead of github.com, since latter is a mirror.
- Include both develop and 6.3-backports branches as versions 
- Include a number of patches specific to QE 6.3, documented in source
- Proper conflicts for HDF5
- Spack package default for variants matches autoconf/automake defaults.
- Other minor fixes

### CODE REVIEW

**Comment by naromero77** [comment]
So 'spack flake8 packages.py' says 'Flake8 checks were clean', but the TravisCI check does not like the escape sequence here below and results in W605 invalid escape sequence
https://github.com/spack/spack/blob/8c0da9cdb50dfb23d7516835e3f7b82d5579543d/var/spack/repos/builtin/packages/quantum-espresso/package.py#L203

Does anyone know how to fix this? I am unsure of what that segment of code actually does.

**Comment by scheibelp** [comment]
> `'HDF5_LIB([\s]*)=([\s\w\-\/.,]*)', `
> Does anyone know how to fix this? I am unsure of what that segment of code actually does.

I think prepending the string with `r` like `r'HDF5_LIB([\s]*)=([\s\w\-\/.,]*)', ` will fix the error (i.e. use a raw string).

I'm not sure if this is something flake recently started complaining about, but I think it would be good to change:

Normally `\` in a string indicates an escape sequence. In raw strings `\` just means the `\` character. Typically regexes that aren't using raw strings have to escape all the `\` characters to make use of the regex character classes (which is to say that I'm suspicious that this regex isn't doing what was intended and that converting to a raw string will fix it).

See also: https://stackoverflow.com/questions/2081640/what-exactly-do-u-and-r-string-flags-do-and-what-are-raw-string-literals

File: var/spack/repos/builtin/packages/quantum-espresso/package.py [hunk 1]
```diff
@@ -17,36 +17,49 @@ class QuantumEspresso(Package):
...
+    depends_on('fftw-api@3')
     depends_on('mpi', when='+mpi')
     depends_on('scalapack', when='+scalapack+mpi')
-
-    depends_on('fftw+mpi', when='+mpi')
-    depends_on('fftw~mpi', when='~mpi')
     depends_on('elpa+openmp', when='+elpa+openmp')
     depends_on('elpa~openmp', when='+elpa~openmp')
-    depends_on('hdf5', when='+hdf5')
+    # Versions of HDF5 prior to 1.8.16 lead to QE runtim errors
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/quantum-espresso/package.py]: typo:
```suggestion
    # Versions of HDF5 prior to 1.8.16 lead to QE runtime errors
```

File: var/spack/repos/builtin/packages/quantum-espresso/package.py [hunk 2]
```diff
@@ -60,30 +73,86 @@ class QuantumEspresso(Package):
...
+    # QE 6.3 requires multiple patches to fix MKL detection
+    # There may still be problems on Mac with MKL detection
+    patch_url = 'https://gitlab.com/QEF/q-e/commit/0796e1b7c55c9361ecb6515a0979280e78865e36.diff'
+    patch_checksum = 'bc8c5b8523156cee002d97dab42a5976dffae20605da485a427b902a236d7e6b'
+    patch(patch_url, sha256=patch_checksum, when='@6.3')
+
+    # QE 6.3 `make install` broken and a patch must be applied
+    patch_url = 'https://gitlab.com/QEF/q-e/commit/88e6558646dbbcfcafa5f3fa758217f6062ab91c.diff'
+    patch_checksum = 'b776890d008e16cca28c31299c62f47de0ba606b900b17cbc27c041f45e564ca'
+    patch(patch_url, sha256=patch_checksum, when='@6.3')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/quantum-espresso/package.py]: Just a heads up, these patches may break if and when a version 6.3.1 is released. See #8320, #8432, #8957, #9281, #9307, and #9330. To future proof this, you can hack it like so:
```suggestion
    patch(patch_url, sha256=patch_checksum, when='@6.3:6.3.0')
```
as was done in #9330.

**Comment by alalazo** [comment]
@naromero77 Thanks! 

@nvarini FYI

**Pull request merged**